bee-network/bee-autopeering/src/peer/mod.rs

```diff
@@ -0,0 +1,334 @@
+    ser::SerializeStruct,
+    Deserialize, Serialize,
+};
+
+use std::{fmt, net::IpAddr};
+
+/// Represents a peer.
+#[derive(Clone)]
+pub struct Peer {
+    peer_id: PeerId,
+    ip_address: IpAddr,
+    services: ServiceMap,
+}
+
+impl Peer {
+    /// Creates a new instance.
+    pub fn new(address: IpAddr, public_key: PublicKey) -> Self {
+        let peer_id = PeerId::from_public_key(public_key);
+
+        Self {
+            peer_id,
+            ip_address: address,
+            services: ServiceMap::default(),
+        }
+    }
+
+    /// Returns the [`PeerId`] of this peer.
+    pub fn peer_id(&self) -> &PeerId {
+        &self.peer_id
+    }
+
+    /// Returns the public key of this peer.
+    pub fn public_key(&self) -> &PublicKey {
+        self.peer_id.public_key()
+    }
+
+    /// Returns the address of this peer.
+    pub fn ip_address(&self) -> IpAddr {
+        self.ip_address
+    }
+
+    /// Returns the port of a service provided by this peer.
+    pub fn port(&self, service_name: impl AsRef<str>) -> Option<u16> {
+        self.services().get(service_name).map(|s| s.port())
+    }
+
+    /// Returns the services of this peer.
+    pub fn services(&self) -> &ServiceMap {
+        &self.services
+    }
+
+    /// Sets the services of this peer all at once.
+    pub(crate) fn set_services(&mut self, services: ServiceMap) {
+        self.services = services;
+    }
+
+    /// Returns whether the peer provides a corresponding service.
+    pub fn has_service(&self, service_name: impl AsRef<str>) -> bool {
+        self.services.get(service_name).is_some()
+    }
+
+    /// Adds a service with address binding to this peer.
+    pub fn add_service(&mut self, service_name: impl ToString, protocol: ServiceProtocol, port: u16) {
+        self.services.insert(service_name.to_string(), protocol, port);
+    }
+
+    /// Returns the address, i.e. the `libp2p_core::Multiaddr`, belonging to the given service name.
+    ///
+    /// Example: "peering" => `/ip4/127.0.0.1/udp/14627`.
+    pub fn get_service_multiaddr(&self, service_name: impl AsRef<str>) -> Option<Multiaddr> {
+        if let Some(service) = self.services.get(service_name) {
+            let mut multiaddr = Multiaddr::empty();
+
+            match self.ip_address {
+                IpAddr::V4(ipv4_addr) => multiaddr.push(Protocol::Ip4(ipv4_addr)),
+                IpAddr::V6(ipv6_addr) => multiaddr.push(Protocol::Ip6(ipv6_addr)),
+            };
+
+            multiaddr.push(service.to_libp2p_protocol());
+            Some(multiaddr)
+        } else {
+            None
+        }
+    }
+
+    /// Creates a peer from its Protobuf representation/encoding.
+    pub fn from_protobuf(bytes: &[u8]) -> Result<Self, DecodeError> {
+        Ok(proto::Peer::decode(bytes)?.into())
+    }
+
+    /// Returns the Protobuf representation of this peer.
+    pub fn to_protobuf(&self) -> Result<BytesMut, EncodeError> {
+        let services: proto::ServiceMap = self.services.clone().into();
```

Comment:
Is there no way to impl the proto `From`s for references to avoid these kind of clones ?

Reply:
That's basically the main problem with ProtocolBuffers. The `proto::Peer` type demands ownership, which we cannot give up (as we still need to keep that peer in our own peer list), hence a clone/copy cannot be prevented. Does that answer the question?

Reply:
I guess so :smile: